### ISSUE TITLE
JSON AST: remove ``is`` prefix, check for ``fallback``

### ISSUE THREAD

**Issue by ekpyron** [ethereum/solidity#4691]
Came up in https://github.com/ethereum/solidity/pull/4663#discussion_r207832105

``isConstructor`` in the JSON AST should be changed to ``constructor``.
We should verify that fallback functions are in the AST.

**Comment by axic** [comment]
@soad003 this may need to be added into remix. Can you have a look?

**Comment by soad003** [comment]
@axic yes i use this property in one module. I could just check both properties for now, to avoid to much coordination. What do you think?

**Comment by axic** [comment]
That would be great to check for both. Do you think this could be done within 2 weeks in Remix?

**Comment by ekpyron** [comment]
Currently, fallback functions just end up as functions with empty name - and actually that should be sufficient or do you think we need some explicit ``fallback`` field or something similar?

**Comment by axic** [comment]
I think better to have that, since we do have that internal helper and other the ABI has a specific flag too.

Then the consumer can assert that name must be empty if fallback is set, and if it isn't that means there was a breaking change.

Happy to take arguments against it.

**Comment by ekpyron** [comment]
Well, the reason for removing ``constant``/``isDeclaredConstant`` and ``payable`` was to remove redundancy and strictly speaking a fallback field would be redundant as well, so that might be an argument against it... Are there any other redundant fields in the AST? We probably don't have to avoid redundancy at all cost, though.

An argument *for* having the field would be that ``name`` is also empty for constructors, so checking for a fallback function would be ``name === "" && !constructor``, which is rather involved.

Another argument *against* it would be that ``\"fallback:\" false`` would probably be added to all functions, increasing the size of the AST - however, the same is true for ``constructor`` as well, so...

So in summary now I tend to agree with you and would add a ``fallback`` field.

**Comment by axic** [comment]
Or: have a `kind` field with `fallback`, `constructor` or `function` (mirroring what is in the ABI JSON).

**Comment by axic** [comment]
This `kind` option might be better if we consider the split of fallback into 2.

**Comment by ekpyron** [comment]
Unfortunately, such a field would somewhat depend on the decision in #3198. I like the ``kind`` field, but for now I would just use ``fallback`` as one kind and rely on the ``payable`` field for further distinction.

**Comment by axic** [comment]
I mean having a `kind` (or `type` is full compatibility with the ABI JSON is wanted) would put is into a forward compatible place: if we decide to split fallback into 2, we wouldn't need to add a new filed and deprecate the current ones. (I do think the split is agreed to, just we do not know how to name them.)

**Comment by ekpyron** [comment]
My last comment was ill-written - I meant to say: adding a ``kind`` field even now is a good idea, but I would use a single entry for both fallback types, i.e. ``fallback`` for payable and non-payable fallbacks (they are still distinguishable by the ``payable`` field). After naming in #3198 is decided, we can change into two names (or not).

**Comment by ekpyron** [comment]
I created PR #4734 for the ``kind`` field.

However, I noticed that the ``is`` prefix is used for in quite a few non-tested cases, which first suggests that we need more extensive tests and secondly we need to decide whether to change all of them.

They are:
  -  ``isInlineArray`` for tuple expressions
  - ``isConstant`` for expressions
  - ``isPure`` for expressions
  - ``isLValue`` for expressions
  - ``isSlot`` for inline assembly identifiers
  - ``isOffset`` for inline assembly identifiers

Also: should we maintain backwards compatibility in the legacy AST and only make further changes to the new AST? That would be relevant for #4734 and for changing the ``is`` cases above, if we want to do that.

**Comment by axic** [comment]
Actually I would prefer to drop support for the legacy AST (hoped that we could do it in 0.5.0, but didn't make any sense, hence it is too late).

**Comment by ekpyron** [comment]
Ok - but since we can't drop it now, should we rather apply changes like this one to both ASTs or leave the legacy AST unchanged? Actually, for what reason is the legacy AST still there? Because of external projects depending on it, I guess?

**Comment by axic** [comment]
I don't think there was any assessment done yet about the users of it, but I think Remix still uses - @soad003 can you please correct me if I'm wrong?

Minimising the number of changes made to the legacy is fine as long as it doesn't require way too much work from our side.

I think for 0.5.0 we should just make a decision on this single PR and do an overhaul after. I'm leaning towards using kind/type, which solves all the questions we have here.

**Comment by ekpyron** [comment]
I agree - I think we should just merge #4734 for now (which removes ``isConstructor`` and adds ``kind``). But changing #4734, so that it only affects the new AST and the legacy AST stays the same, is trivial, so I'll just do that (I'll use a separate commit, so we still have both options, in case anyone comes up with arguments for either option).

**Comment by ekpyron** [comment]
I just adjusted #4734. If we still want to do the change in both ASTs, we can just remove the last commit from it.

**Comment by ekpyron** [comment]
An argument for changing it in both ASTs would be that we did the same for ``constant`` and ``payable`` as well.

**Comment by soad003** [comment]
@axic yes I still use the legacy AST, are there any resources concerning the new ast (what changed, etc.)? Such that i can at least judge how much effort it would be to switch. But i kind of suspect that it is to much for me to handle at the moment. But i am happy to give it a look.

**Comment by ekpyron** [comment]
@soad003 I don't think the JSON ASTs are all that well documented. I created issue #4758 for that. There is some examples of both legacy and new ASTs in https://github.com/ethereum/solidity/tree/develop/test/libsolidity/ASTJSON, but I think so far they are far from extensive. So I think it probably makes more sense for you to only look into it after we extended documentation and tests - that probably won't happen before 0.5.1, though, and we will probably have to keep the legacy AST until 0.6.0.

**Comment by soad003** [comment]
@ekpyron Thanks for creating the issue. It just skimmed through the examples really quick but it is hard parse the difference from there alone. As far as i could parse some field names changed e.g. children -> node, name (on node level) -> nodeType etc. And the new AST got rid of the nested attributes object, the properties are now included in the node itself. A high level description of the changes like that would be helpful.
